plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Parse/Comments.hs

```diff
@@ -0,0 +1,576 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections #-}
+
+module Ide.Plugin.Eval.Parse.Comments where
+
+import qualified Control.Applicative.Combinators.NonEmpty as NE
+import Control.Arrow (first, (&&&), (>>>))
+import Control.Lens (lensField, lensRules, view, (.~), (^.))
+import Control.Lens.Extras (is)
+import Control.Lens.TH (makeLensesWith, makePrisms, mappingNamer)
+import Control.Monad (guard, void, when)
+import Control.Monad.Combinators ()
+import Control.Monad.Reader (ask)
+import Control.Monad.Trans.Reader (Reader, runReader)
+import qualified Data.Char as C
+import qualified Data.DList as DL
+import qualified Data.Foldable as F
+import Data.Function ((&))
+import Data.Functor.Identity
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import qualified Data.List.NonEmpty as NE
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Void (Void)
+import Development.IDE (Position, Range (Range))
+import Development.IDE.Types.Location (Position (..))
+import GHC.Generics
+import Ide.Plugin.Eval.Types
+import Language.Haskell.LSP.Types.Lens
+    ( character,
+      end,
+      line,
+      start,
+    )
+import Text.Megaparsec
+import qualified Text.Megaparsec as P
+import Text.Megaparsec.Char
+    ( alphaNumChar,
+      char,
+      eol,
+      hspace,
+      letterChar,
+    )
+import Data.Functor ((<&>))
+import qualified Data.Text as T
+
+{- |
+We build parsers combining the following three kinds of them:
+
+    *   Line parser - paring a single line into an input,
+        works both for line- and block-comments.
+        A line should be a proper content of lines contained in comment:
+        doesn't include starting @--@ and @{\-@ and no ending @-\}@
+
+    *   Line comment group parser: parses a contiguous group of
+        tuples of position and line comment into sections of line comments.
+        Each input MUST start with @--@.
+
+    *   Block comment parser: Parsing entire block comment into sections.
+        Input must be surrounded by @{\-@ and @-\}@.
+-}
+type Parser inputs = Parsec Void inputs
```

Comment:
It looks like this function is not reused anywhere except `LineGroupParser` definition and `parseLine`. Do we need this?

Reply:
I suppose you are talking about 'Parser` *synonym*, right?
If so, yes, this could be eliminated. Thanks!

Reply:
Sorry, yes. I meant the type synonym.